### ISSUE TITLE
Type logical operators "and", "or" and "not" in extends clauses for mapped types

### ISSUE THREAD

**Issue by inad9300** [microsoft/TypeScript#31579]
## Search Terms

Pretty much the ones in the title...

## Suggestion

I would like the ability to use logical operators for types.

I am aware that `not` is already planned as a feature, but I would like to make sure that this gets further extended to other logical operators, and I couldn't find any hints that this is in your minds already.

## Use Cases

Type composition, better readability...

## Examples

Current tricks to get to the desired behavior:

```ts
type Not<T extends boolean> = T extends true ? false : true

type Or<A extends boolean, B extends boolean> = A extends true
    ? true
    : B extends true
        ? true
        : false

type Or3<A extends boolean, B extends boolean, C extends boolean> = Or<A, Or<B, C>>

type And<A extends boolean, B extends boolean> = A extends true
    ? B extends true
        ? true
        : false
    : false
```

A few arbitrary use cases:

```ts
type Primitive = boolean | number | string | symbol | null | undefined | void
type IsA<T, E> = T extends E ? true : false

type IsIndexSignature<P> = Or<IsA<string, P>, IsA<number, P>>

type IsCallback<F extends Function> = F extends (...args: any[]) => any
    ? And<Not<IsA<Parameters<F>[0], Primitive>>, IsA<Parameters<F>[0], Event>> extends true
        ? true
        : false
    : false
```

All together in the Playground: [here](https://www.typescriptlang.org/play/#src=type%20Not<T%20extends%20boolean>%20%3D%20T%20extends%20true%20%3F%20false%20%3A%20true%0D%0A%0D%0Atype%20Or<A%20extends%20boolean%2C%20B%20extends%20boolean>%20%3D%20A%20extends%20true%0D%0A%20%20%20%20%3F%20true%0D%0A%20%20%20%20%3A%20B%20extends%20true%0D%0A%20%20%20%20%20%20%20%20%3F%20true%0D%0A%20%20%20%20%20%20%20%20%3A%20false%0D%0A%0D%0Atype%20Or3<A%20extends%20boolean%2C%20B%20extends%20boolean%2C%20C%20extends%20boolean>%20%3D%20Or<A%2C%20Or<B%2C%20C>>%0D%0A%0D%0Atype%20And<A%20extends%20boolean%2C%20B%20extends%20boolean>%20%3D%20A%20extends%20true%0D%0A%20%20%20%20%3F%20B%20extends%20true%0D%0A%20%20%20%20%20%20%20%20%3F%20true%0D%0A%20%20%20%20%20%20%20%20%3A%20false%0D%0A%20%20%20%20%3A%20false%0D%0A%0D%0Atype%20Primitive%20%3D%20boolean%20%7C%20number%20%7C%20string%20%7C%20symbol%20%7C%20null%20%7C%20undefined%20%7C%20void%0D%0Atype%20IsA<T%2C%20E>%20%3D%20T%20extends%20E%20%3F%20true%20%3A%20false%0D%0A%0D%0Atype%20IsIndexSignature<P>%20%3D%20Or<IsA<string%2C%20P>%2C%20IsA<number%2C%20P>>%0D%0A%0D%0Atype%20IsCallback<F%20extends%20Function>%20%3D%20F%20extends%20(...args%3A%20any%5B%5D)%20%3D>%20any%0D%0A%20%20%20%20%3F%20And<Not<IsA<Parameters<F>%5B0%5D%2C%20Primitive>>%2C%20IsA<Parameters<F>%5B0%5D%2C%20Event>>%20extends%20true%0D%0A%20%20%20%20%20%20%20%20%3F%20true%0D%0A%20%20%20%20%20%20%20%20%3A%20false%0D%0A%20%20%20%20%3A%20false%0D%0A)

Desired syntactic sugar to write the same:

```ts
type IsIndexSignature<P> = IsA<string, P> or IsA<number, P>

type IsCallback<F extends Function> = F extends (...args: any[]) => any
    ? not IsA<Parameters<F>[0], Primitive> and IsA<Parameters<F>[0], Event> extends true
        ? true
        : false
    : false
```

It would make the most sense to accompany this with better support for boolean checks in type definitions. That is, to allow to use the ternary operator directly without the constant need for `extends true` everywhere. Possibly, a new sort of "boolean type declaration" could be introduced, as to avoid having to propagate the boolean value all the way. For example, it should be possible to define [KnownKeys](https://stackoverflow.com/questions/51465182/typescript-remove-index-signature-using-mapped-types) (not full implementation here) like this:

```ts
type KnownKeys<T> = {
    [P in keyof T]: IsIndexSignature<P> ? never : T[P]
}
```

Without the need to do:
```ts
type KnownKeys<T> = {
    [P in keyof T]: IsIndexSignature<P> extends true ? never : T[P]
}
```

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by MartinJohns** [comment]
I don't think introducing more keywords and additional syntax for something that works well with existing tools already is worthwhile.

**Comment by inad9300** [comment]
Except it doesn't work well... It is hard to write, hard to read, error prone, and at the same time something quite fundamental, which projects will keep duplicating on and on.

**Comment by jcalz** [comment]
I've more or less switched to using `unknown` and `never` instead of `true` and `false` for this sort of thing.  It has the added benefit of letting you use `&` for "and" and `|` for "or".  If `not` ever lands, you can use that; for now you need a generic type alias like `type Not<T> = [T] extends [never] ? unknown : never` or maybe `type Not<T> = unknown extends T ? never : unknown` (depending on what you want to see happen with `Not<123>`)   

After that the only thing missing is the abbreviated conditional type syntax `C ? Y : N` which stands for something like `[C] extends [never] ? N : Y` or maybe `unknown extends C ? Y : N` (depending on what you want to see happen with `123 ? Y : N`).  I wouldn't hold my breath waiting for that to make it into the language, so I guess the best you can do there is a generic type alias also: `type Cond<C, Y=unknown, N=never> = [C] extends [never] ? N : Y`.

**Comment by inad9300** [comment]
@jcalz could you illustrate your ideas by applying them to the examples above? For instance, to `IsIndexSignature`, perhaps explaining it a bit? It is quite tricky to become certain that `|` will equate to a logical by somehow using `never` and `unknown`.

I am curious, because I want to have such a feature today if this is possible, but it is most definitely a hack that I don't wish anybody to be forced to use for what would have otherwise been a very easy task.

**Comment by jcalz** [comment]
I suppose "hack" is in the eye of the beholder, here.  While the boolean literal types `true` and `false` have evocative names, they just represent the types of boolean *expressions* that exist at runtime, they don't behave like booleans *at the type level*.

On the other hand, [`unknown`](https://github.com/Microsoft/TypeScript/wiki/What's-new-in-TypeScript#new-unknown-top-type) and [`never`](https://github.com/Microsoft/TypeScript/wiki/What's-new-in-TypeScript#the-never-type) are TypeScript's names for the [top](https://en.wikipedia.org/wiki/Top_type) and [bottom](https://en.wikipedia.org/wiki/Bottom_type) types, respectively, which behave much more like booleans at the type level, which is what you're asking for. 

The correspondence  comes from equating a type to the statement that some value can be assigned to the type:

| Propositional Logic | Type System |Correspondence
|-------|-------------|-----|
|  T/true/⊤    | `unknown`  (top)           | value can be assigned to `unknown`?  *true*. |
|   F/false/⊥    | `never`   (bottom)          | value can be assigned to `never`? *false*. |
| and/∧      | `&`  (intersection)           | value can be assigned to `A & B` iff it can to `A` *and* `B`. |
| or/∨ | `\|` (union)  | value can be assigned to `A | B` iff it can to either `A` *or* `B`. |
| not/¬ | [`not`](https://github.com/Microsoft/TypeScript/pull/29317)? (complement) | value can be assigned to `not A` iff it is *not* to `A`. |

Hopefully that seems less arbitrary.  Of course, there is no `not` yet, and propositional logic doesn't usually have a primitive symbol for a ternary operator, so we'll need our own `Not<T>` and `Cond<T,Y,N>`, as I said before:

---

So, let's look at your example:

```ts
type Primitive = boolean | number | string | symbol | null | undefined | void
type IsA<T, E> = T extends E ? unknown : never;

type Cond<C, Y=unknown, N=never> = [C] extends [never] ? N : Y
type IsIndexSignature<P> = IsA<string, P> | IsA<number, P>

type Yes = IsIndexSignature<number> // (never | unknown) = unknown
type No = IsIndexSignature<boolean> // (never | never) = never
```

Does that make more sense?

**Comment by inad9300** [comment]
Hmm... Sort of. I have tried to replace in a bigger example more or less every occurrence of `true` with `unknown`, `false` with `never`, `Or<T>` with `|` and so on, and I failed miserably. The thing is, previously `IsIndexSignature<T>` was easily testable for use in bigger types via `extends true ?`. Now this flexibility is lost, and I am (I think) forced to test with `extends never ?` (or `extends [never] ?`? I didn't really get why wrapping in an array is necessary...), since everything extends `unknown`, and thus reverse all the conditionals (quite a few).

The bottom line is, I do not believe your proposal is a reasonable one. *Hack** is not so much in the eye of the beholder, I would say. This arrangement of tricks is by any metric harder to understand than "regular binary logic" -- in my opinion, to the point where the pain is too high and making mistakes is significantly easier.

**Comment by fatcerberus** [comment]
> I didn't really get why wrapping in an array is necessary...

If you mean this pattern: `[T] extends [U] ? Y : N` - it's to turn off [union distribution](https://www.typescriptlang.org/docs/handbook/advanced-types.html) (see section **Distributive conditional types**).  Generic conditional types map over the individual members of a union so if we don't want that to happen, we turn the union of types into a "tuple-of-union" which is treated as one atomic type.  You could replace `[T]` which pretty much any other covariant generic and it would work the same way.

**Comment by yume-chan** [comment]
It's sly that TypeScript Handbook describes Conditional Types as

> A conditional type selects one of two possible types based on a condition expressed as a type relationship test:
> ````
> T extends U ? X : Y
> ````
> The type above means when T is assignable to U the type is X, otherwise the type is Y.

(from https://www.typescriptlang.org/docs/handbook/advanced-types.html#conditional-types)

It never says Conditional Types will resolve to `true` or `false`, so if we want to make specific the result of the condition, any binary values can work, `unknown` and `never` are definitely ok.

---

#### But not `true`/`false`

Everyone knows that at runtime the opposite of `true` is `false`, and vice versa (except NaN, of course).

But in negated types, `not true` means "anything not assignable to `true`" and `not false` means "anything not assignable to `false`".

So for example `string` is `not true` and `not false` at the same time, what will `string ? true: false` resolve to?

---

However I have to admit that using `unknown` and `never` is way more less intuitive.

---

Originally I came here because I found a super cool [blog post](https://zhuanlan.zhihu.com/p/89571439) (in Chinese) that uses only the type system to build a script parser.

![image](https://user-images.githubusercontent.com/1330321/68039823-ecbe3e80-fd07-11e9-9f22-026fd9fe0f40.png)
([Playground link](http://www.typescriptlang.org/play/index.html?ssl=356&ssc=1&pln=352&pc=1#code/PTAEgX4x070Mm9BiVRnRUHSph1bUFj-BYAUFgLgTwAcBTUAQwBNzQBeUAIgvLoG5Q2RRACpV0JIGcArgCMa9QUJZtQHQPOKPYqAC2AgDai6ylZPZhAGnLyS5AJYA3dcZPbpYQCZpBpQHsqtDU6sdAa8qBNeXtEAjur+VjqggCRKgLWmgB9ugJLe9gB2JC4JwdZh4YCwcrGY+AoqOOp5KRwRgPA6gHFy9iqJ9FVFYKVl0RXZvKAA5vkuHXVpgOSazTkkbdV0wz0R-U32AJIAZupzLPYAoip8I6vrSy0Km0Rz6nuLzPYASkQ4AgBOcernlzfbWByADqaAI36AsyaAOvIY2JgcgHbgjyAGm9AABygCo5QDCioACX3sAAkiBQADwAFQAfKIUaAiAAPHBEOLkPigADeoAADAAuUBGOKzIhXUBw0AAX1AAH4maBqQkTAyTv8wIBvzxsYKhsJ2JBRpCMKlRGNoAAosFJQIqAHSa0jUrG4-GE4mkOJ4ADaAF0OaAsTyiHyrgBKGgYkwOIzkLCOvUEolq7VkY0AGlAmvVQmptPpjNOjuoztdVE5p25oF5-OeYEAFOqAFk1ABcJgCXjQCmihCYfYAKpxPgACyMsxwqKDAEEFWqVVJFX6G0GQ2Grdi8d7DcbzZbrSnbQyY3G3R6+-qfRqteG6QzQNGnaAXW7LUmbXaBRwM4A+HRznCL4vsAGEHAQ8PXQABlWcD-14Ztkk0ABRptwA1kQ8A48woma1L3p+FosgKgygB+AhVneTaiFeN5IuWVY1nWKJBkar5BliABkq5EAAxg4VzkEifA4FctJtI2aJogK6agIAYEqAAnmgAWaoAUHKANBygDhpvYH6kFc6xXFecT4niqIOH+FaiDhQb3Ncty0Dhb6tmwOAyQSfA6tpFasKqHCAMnxBaACFuoBabJGmgFcFzKdSSk3IZUgcIAK-GAHtqgDYSoAX3pmYAQUGAJ0OWAslgzGABSugCm5oAFhGAFyegCLyoA+K6ALfRgCrNoAmKmAPfR9j3ngEmkDiyynKcADypyoqIJIsupmCqnZFAOHEKh4A+eU4AVyxXFcZHUgIcRGH4Aj8HgihCA4KgCqqihEHwfCkMMOoCqFfwcIAYDqeYAPAqAH3RgB2-oAXHKAKDKgDUKoAYXJZYAh-KAPYGZlpVlgDePoA0er2CilbdQA7rl+WFV1ZFIuJkk4EGyw4sQxE4AqNn-XqT4GjBwmiVDUkRiuKL6Xw2HGmiNmcveFxOXEf2NQDQafe1hXFWVSIkjZqp0OWuKg-iVBWQSdDUh+RBED+qOyYTEl6oxNNSHQwOM2zoCiyROCTaqwsNjg4s8zpMtsNVWO1VIu5pitYCAHNyoCI-kHiAOn6tn2TcoBmRt9gAOJ4+bBOGxDGtsIbMM+kJIkMobSIKd+karhiibJqmVxMYKoCAJymgAbymbDy3BtgAQ-4AedqAKJpXGAPRmnkQoADaZmbxfHJ4A0fKAJD-OX2-HfPE0R8fO6qbterDnsI0TepIsjjJKxWQYd6AAAypBUerRlgIAQZqADnmgBPuoA835JVlgCMOukgCxioAd-KR5tJ3Y-3g-5I3Pqkx1FNJpyhs2ez8Pe63Uld+jNfKRiWth2FEfIIARumAPOJj32H3Rj4lcpBrFEHEAQo0VwAB9QBjXGoiOI4cOBQGgPYAAsqoHARgCAqCMCVL8tBNCgAgRYfBjhyBQVaA2Sgv9TBEGwfJSgRDxCkIUFeRQBBhJGD4I1GhtB-BEISEQvI-CSAQI6EQ4YcCwCAAtFHMSBkD2AANL-jemRH0tADgQL2EQo48wIH4yISgvI6DME0IgeQ4waC+TGINg4FhbCOFxGweI0AJseIQksvpS2X9JQwU5tzfSVcBY0EhlfXe-Ym4XzEsE9uy5O5ox7tEu+Nxh5SETA7d2xID7k1KsfegRUsl0DPlaNGaTQDDk5CLEqAAxfJLtkw32KWSKkoAFF4CUeRYkbJOQ3xNOSM0BS6l72JA06kVEaJxDaKyS0dA3QEjQbMIwDJqmqj0rJYkAzSQUmpD-P+AD2mTMwdsrQBS6B9R-HEBwb04h0GTMcuIpzzmXMcYAb+iPAuPBG46yXj7zESNP4vEQMQZSzrlIBuoSPbhJ9r3G+cSA74ySSEGec9Mrz0AC9ugBS408lvXRayMm5LKpaU+NSOAT0eoATmVACyiYAGH+MwF0ALRynlACWSl4MyplYAmwsizOIfSilrKGRLAFYNWRb2bpffm18ZRyhvmiPCaNulmjhakdI5KKWAGyjQAX4pZT6a9c5OKfpXF+YDPljN5WP0cYAUPjACVSp46C943q-2IpWO8pZilqUCTU3UoLiR-gAvMJ1nJSwmmAsmf1dByBEFmKQVBdAzSOLSkFQAX+qeNpH-cNxESAZIAHJOBINTGpgxhnUVoktZ+SaGQppIHMe87V8TTQknzMxRhGp4UrASIGawiAYmxW1AqmbQ2khsnm+gFaq1EBrQrFWoBSKEkoY1ak4l62NXHTgZtcQdTLvHUQNt1I9hFr+CWq4ZaElxEraQatMykSizsrNBtcQO0etal9Ht2b+28GpHQfGx7T0SW2KqBml6+DXq3SDP916d3MTfu-aKgA++MAPD6WA90Hq2QyABF6Zr-sakiAAagAoat65zpK7TiR9fbc0vvoIh-+KgUNXsat+qQJhsNEGpFhlQQ0XIcHA9bTAy1mKAHALQAY9HQCg7BzA8HSCptANMUNEkazzKuFRtDBN02kGmrh58Gas3EdVAOugkmZkyYZPJ69tG2BxGU4x0ASnpouRCPx6AgBYFUAA+eIVn4cEADFykIhNwf5vusTJAABCtJhJ4EM+h7BQY+5wnvIpSLqnYbqd7TmrTpG6ABdM1cYLQHUNGfHdeakDjZYjwgDAGyKhKy6X7pF6zrkwCADPdTzNSrhlccpVgrqRADPyvV7jEd2KgkAIbmyVACnyoJC+H6R1nqdq61UNq7UOo5lzLu+qpWada3Mc+Xsh0nrG7Wp247VT4zW+sd9w7R2Ld21IUN4bUEHaICFx2wTBY1OquHaCwqNufrrIbaVslilqIlm2g4qjZhArYF8n5n3CmyViz6XuBLCvsV64AN7l6uqk5MK27i3ikw+CVvO2OB8YY7WVjqdaDGo2Q4OxFigB6U0G4AdXUhUXz8yoBwxEfwE7vUTgGOOK7KTZ3h-2KNl1k7AOxQAboqADW5Kn9OvZ7E9h9+7mP4mw+SaAXH+OJuE-iXsIXoBPLsVVQFYKNScbc5uPqoMb2tty+J9eptLa-vrAYgU0OIdxyMkfi7u0HuVzu5Nc-F7F8ZfCStwDYHD5bU4HtUiObvjeZOyDIl1rexruM+Z6z0Hd3RUGr2AxM7bAtHXYtyd8HWiHutdABdiNeRQIm4z9XYBKgVCl6kE9v3rRhVHc20X+X1AbLp7N4eqHxIOd6il+sdH6v2eK+x0blXNfefPl7rdreuM8cO37x39757MvUZvcascnuff7+1sxdi9X-fS+3wpjHPeanCuYawmidjx-3ccexONgAaczP23i+H4aKKCC8-pnqHtNhHrNj4gtnHstgVtMtJnMgyAXlJrMrJoASHrnqAPskhioNduRshpfteqdkLGwBXldlaFqh9ARp1N1HquDnQL+lljRk3qrA9ifmxLZl-goK9ogfpnJngehhNjflNt8rXnqEGCMrRIPvzoyErmwJyKrmvhPnzr3JZkQMvnPuDrprAcgbwYpmZo7jUofqHK-mxOBuwSQMKjgZRtodfr3kIf3hYXwBIcPniFzqvpXAoQvvEsxkNKoa4TzuDhYejl4e2nvs7r7jrKxGxIAC9mgAG1mAAY8oABKKgA5X5mSmFwxexwhGBtDNpXC-4No0T4AXr97YKh40JrL362KcJfio4XymKUJ8goECwFKlF3q1HmLUJVFpHrD6JoIYJGANF4h77NF87dGGJYIdHCq5H-7pb9HgxO6u5GG6CADkBggqkcKiMb0TMaHsKqlgAVYYQZ0UQJMbsQQMBnwfdgGPsesUYh+DZMwd1mxJwMsTAKsTURQm0ZsZNlINsYFtMXsTUqqGsagqMZsRcf8VIK0VQtgrcUYYACvWgAwfErEwbDZezlGP6NQfECFfEM4-EZYnH0EEz7HCoQn1FWFx77GonsKVHQkubC5sTuZCYUrAiubgiABDyi8IAODGgAWdqACBkexGZAgsiesDsb8XiTvv3pLGDMUZVDZI0obKwBwOkBnLyWxPyTABSmihin8IVibDVoAJ3xKRSJNSAAjNSHIW4fLhrgHBFveNrs8gKTPn3sXvyjgFKZaSuNIZaGaX4RaZPgHFCYSmAG1rqYACl6gAp9EZwxEJHJHI7K7CoZFZEMi5FkS-y4nUHBJBium+nunT4o6z6+Gm7uGwy9ynCRba6eRCaABueoAOd+gAt36ABV+oAPXOo+RAwpqZ+JSI+xK+au6ZoArZ6OYWFWUWq4MWUq+xEpLpNxNSIRruXubuR+c5oR85S0Jo2uEQry7yBIoAGcipypqpiCt+4BfihZPo45NCnIRpyY5IWA0aNJoAlYYmP4gpRAo2o6fARp1hB582R5PpfOIsR8VyEC5SVS+KOZmseZ3ZQBCuAcL5MyP8VE2uPW-WQ2M+wqMF22P5HhAcHpshahGFRZ8SaFOA2uBAcElYoAgA2-GDbvAVmam5ldnyE9mwTwSEVwUGqEV6FLLzmzmznPbf5eyEV8AABMH5AJI2x2Myb58++FWFoFMhBxAl75E2YRT5AlAALCJVifxeJRJEJVJdDlPpns2QJcJUpUua3hwWJZ3hJQABwaVsCoXaU4B8DqXHlD4GWc4oWWXvbOUY7KVeIOVWU6VGkABsdl8ljlfAtlrlkhVihlnlWlgVTlUV8ufl5+6wAlAAzCZd3jZAFd5SFXpW5TJXFbmXlZbm+aFaZYYeZWYV5eVdflAaACabFQDOOo0mVa+VHnVa+VlYtg9iyJ+IebHnhT6H+XkkQkBVcueZeTeeEbrKZOyu4uxGlIAIgWT50kKyzqQ4cqnxbAXppuFSqgKgwqG1OkOet5gAMdoSI9aACt1qZPYIdQ3idZKrtTBHZM9WjI4e5SPihQ4FRCdVVfMbeYACEZ8R7EpkNKN0A2dO-l71F8p1FYNUqon434oAXqgEEOOkIEmNFY4EW1povSAZoAE8gAyvqACyShubcBnO8BSstdDVvI9cdfDV0h+HKtruLpTaAOkIAFoKgAIJqABgLlyisuBKBhHIAHAqNZouTKg2gkf1OAH1KyWx3VElXVXsESmeCNt85o514Ra8oIPEtOZkycgADc5Pkp4s5hUmTmS07TyADj8aDexGGdHIAAhGJ0x0gAy37jz5z8SACf2oAIYxNhYOPZLqbpUhslnpuFkFodvZTOLO2FMdqeIKfOwq6tHlhWnkDtbE+uhtntPhEF1c+1BM5tP4Tse+L070OqVB-edApEvaQgsdP4dA05B+Zl4RgA0raACr0exE1eSOSKAL1oAAHegAqsoy1eL-7GgCWiAnUrnGlBiCVBgZVBiqVBgACsQYwVQYAA7EGNZUGAAJxBhGnkigmqhGlz0L1L2r3r1b07372H3H02Rn2gDz2gCL2gDL2gBr2gAb2gDb2gC72gAH090n1SBP0v1v0f1f0-1-0ANANH0gNsBgMX3v1X3f033-133AOP3n2v2X2f3X2-232AP30INNU4MQOoPQNENwMP2z3P3IOQMEMwOYPwPYP0O4MoP4NoOEMYPENYN0PgN4NQPoOwMkPXl3EcCd3sSABYmoAEI6o90EoaRABAAAIso0ulPfDTPaqHMEGNRENEGNoytrMHo1cAYyUvsToyY5ZGY0QIY5Y8Y6Y+Y0Y2Xq4xJtY-o3YxY2CW46qITb47LP4740EwViE6AITWzX8NBDvJo17EiC4+40GHQHgHQEGB0EkziKk942XvjEkykzZGaKQ7k-QJk+I8wEAA))

To chain multiple expressions, the author has to heavily use Conditional Types. So I wonder is there any proposal to add logical operators to the Conditional Types, as in JavaScript we can use `&&`.
